src/Parsing/Impl/Ast/ScopeInfo.cs

```diff
@@ -0,0 +1,297 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.Python.Core;
+using Microsoft.Python.Parsing;
+
+namespace Microsoft.Python.Parsing.Ast {
+    public abstract class ScopeInfo {
+        // Storing variables due to "exec" or call to dir, locals, eval, vars...
+
+        // list of variables accessed from outer scopes
+        private List<PythonVariable> _freeVars;
+
+        // global variables accessed from this scope
+        private List<string> _globalVars;
+
+        // variables accessed from nested scopes
+        private List<string> _cellVars;
+
+        // variables declared as nonlocal within this scope
+        private List<NameExpression> _nonLocalVars;
+
+        // names of all variables referenced, null after binding completes
+        private Dictionary<string, List<PythonReference>> _references;
+
+        protected ScopeInfo(IScopeNode node) {
+            Node = node;
+        }
+
+        public string Name => Node.Name;
+
+        protected IScopeNode Node { get; }
+
+        internal bool ContainsImportStar { get; set; }
+        internal bool ContainsExceptionHandling { get; set; }
+
+        internal bool ContainsUnqualifiedExec { get; set; }
+
+        /// <summary>
+        /// True if this scope accesses a variable from an outer scope.
+        /// </summary>
+        public bool IsClosure => FreeVariables != null && FreeVariables.Count > 0;
+
+        /// <summary>
+        /// True if an inner scope is accessing a variable defined in this scope.
+        /// </summary>
+        public bool ContainsNestedFreeVariables { get; set; }
+
+        /// <summary>
+        /// True if we are forcing the creation of a dictionary for storing locals.
+        /// 
+        /// This occurs for calls to locals(), dir(), vars(), unqualified exec, and
+        /// from ... import *.
+        /// </summary>
+        public bool NeedsLocalsDictionary { get; set; }
+
+        /// <summary>
+        /// True if variables can be set in a late bound fashion that we don't
+        /// know about at code gen time - for example via from fob import *.
+        /// 
+        /// This is tracked independently of the ContainsUnqualifiedExec/NeedsLocalsDictionary
+        /// </summary>
+        internal virtual bool HasLateBoundVariableSets { get; set; }
+
+        internal Dictionary<string, PythonVariable> Variables { get; private set; }
+
+        /// <summary>
+        /// Gets the variables for this scope.
+        /// </summary>
+        public ICollection<PythonVariable> ScopeVariables => Variables?.Values ?? Array.Empty<PythonVariable>() as ICollection<PythonVariable>;
```

Comment:
Since `Variables` is read/write, attempt to enumerate the collection may throw. Therefore it should be `Variables?.Values.ToArray()`. Also, You can change `ICollection` to `ImmutableArray<>` which is preferable. There is `ToImmutableArray()` extension.